src/passes/OptimizeAddedConstants.cpp

```diff
@@ -123,12 +126,29 @@ template<typename P, typename T> class MemoryAccessOptimizer {
       // code may know that is valid, even if we can't. Only handle the
       // obviously valid case where an overflow can't occur.
       auto* c = curr->ptr->template cast<Const>();
-      uint32_t base = c->value.geti32();
-      uint32_t offset = curr->offset;
-      if (uint64_t(base) + uint64_t(offset) < (uint64_t(1) << 32)) {
-        c->value = c->value.add(Literal(uint32_t(curr->offset)));
-        curr->offset = 0;
-      }
+
+      auto optIfSatisfied = [&](Const* c) {
+        if (memory64) {
+          uint64_t base = c->value.geti64();
+          uint64_t offset = curr->offset;
+
+          uint64_t max = std::numeric_limits<uint64_t>::max();
+          bool overflow = (base > max - offset) || (offset > max - base);
+          if (!overflow) {
+            c->value = c->value.add(Literal(offset));
+            curr->offset = 0;
+          }
+        } else {
+          uint32_t base = c->value.geti32();
+          uint32_t offset = curr->offset;
+          if (uint64_t(base) + uint64_t(offset) < (uint64_t(1) << 32)) {
+            c->value = c->value.add(Literal(uint32_t(curr->offset)));
+            curr->offset = 0;
+          }
+        }
+      };
+
+      optIfSatisfied(c);
```

Comment:
I don't think we need to add this lambda function? The code inside it can just be here normally, without being wrapped in the lambda.

Reply:
address comment, thanks